### PR TITLE
feat: add a program matcher

### LINKED ISSUE #100: Matching on program name but not path
I've been really enjoying this, but I have had a small pain point: the name must be fully specified. Here's an example:

```python
cmake_path = Path(cmake.CMAKE_BIN_DIR) / "cmake"
fp.register([os.fspath(cmake_path), "--version"], stdout="3.15.0")
```

I tried the nuclear option, but it didn't seem to work (and really isn't what I would normally want to do anyway):

```python
fp.register([fp.any(), "--version"], stdout="3.15.0")
```

A really handy shortcut would be something like this:

```python
fp.register([fp.program("make"), "--version"], stdout="3.15.0")
```

which would only match the final component of the path. Maybe this could even be automatic, such as `"thing"` matching `"/any/thing"` or `shutil.which("thing")` it it's not an absolute path. I'd think the automatic one should require it be on the path - but I'm also interested in a fuzzy one that doesn't require it be on the path.

### PR DESCRIPTION
**Pull request by henryiii**
Closes #100.

This adds `fp.program("name")` to match a program regardless of where it is on the path.

Two possible concerns/changes:

* Should this require that the program directory be in the `PATH`, rather than matching for all directories?
* Should this do case insensitive comparisons on macOS if macOS is set to case insensitive (the default)? Is there some existing handling for this I could leverage, if so?

I also had to tweak the types a bit for mypy 0.99x. I personally like adding mypy to pre-commit and having pre-commit handle pinning and updates.

### CODE REVIEW

File: pytest_subprocess/utils.py [hunk 1]
```diff
@@ -127,7 +130,34 @@ def __init__(self, *, min: Optional[int] = None, max: Optional[int] = None) -> N
...
+
+
+class Program:
+    """Specifies the name of the final program to be executed."""
+
+    def __init__(self, program: str) -> None:
+        self.program: str = program
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({self.program!r})"
```
Comment (aklajnert) [inline pytest_subprocess/utils.py:146]: can you implement `__repr__()` and `__str__()` as it is in `Any` class?

**Review by aklajnert — CHANGES_REQUESTED** [review]
Thank you for your contribution, I think it is very useful. Can you please merge with master, so the CI will pass? Also, if you could write one paragraph about this feature in `README.md` that would be great, but I can also do that later if you want.

File: pytest_subprocess/utils.py (hunk 1 above)
Comment (henryiii) [reply pytest_subprocess/utils.py:146]: You mean with the space? Oh, I see, you implemented `repr` in terms of `str`. Any reason why? Python's default `__str__` calls `__repr__`, so if you want them the same, you can just implement `__repr__`. Plus this looks like a repr. Happy to change if you want it changed, just wondering if you have a special reason (like subclasses or something).

**Comment by henryiii** [comment]
If you check "always suggest updating" or soothing like that in the main settings page of the repo, GitHub will add a little "merge or rebase with target" button to the PR, so contributors don't have to open a terminal to rebase on the target branch. ;)

**Comment by henryiii** [comment]
Any comment on my two points in the PR description? Is the simple thing best, or should I change either of those two things? (only match if on the path & case insensitive comparisons)

File: pytest_subprocess/utils.py (hunk 1 above)
Comment (aklajnert) [reply pytest_subprocess/utils.py]: Oh, it seems you're right. There was no special reason, I just thought you need to implement both. Now I know that I don't. So, no need to change that in your code.

**Comment by aklajnert** [comment]
> If you check "always suggest updating" or soothing like that in the main settings page of the repo, GitHub will add a little "merge or rebase with target" button to the PR, so contributors don't have to open a terminal to rebase on the target branch. ;)

Thanks, I didn't know about it. I have enabled this setting.



> Any comment on my two points in the PR description? Is the simple thing best, or should I change either of those two things? (only match if on the path & case insensitive comparisons)

I think the behavior is OK. Checking if the program is in `PATH` could be annoying to the users and require them to use `pyfakefs` in some environments. I'm honestly not sure about this case sensitivity on Mac OS - I would also leave it as is to keep the code relatively simple, as this is a quite unlikely edge case to actually bother anyone.

**Pull request merged**